### DISCUSSION TITLE
Listing packages I've specifically requested

### DISCUSSION THREAD

**Discussion by l008com** [Homebrew/discussions#2346] [Everyday usage]
Hi, odd question. Is there a way to list all of the packages I have specifically, manually requested to install with homebrew, NOT including their dependencies?

**Comment by gromgit** [comment]
I think the following should work (requires `jq`, but that's a easy `brew install`):
```
brew info --json --installed | jq -r '.[] | select(.installed[].installed_on_request) | .full_name' | sort -u
```
The `sort -u` at the end isn't strictly necessary, but helps to weed out duplicates if you have older versions of formulae that haven't been cleaned up yet, or `*.reinstall` Cellar directories from failed upgrades.

**Comment by carlocab** [comment]
If you don't want to install `jq`, you can also try the following script:
```ruby
require "formula"

Formula.installed.select do |f|
  keg = Keg.new(f.any_installed_prefix)
  tab = Tab.for_keg(keg)

  tab.installed_on_request
end.map { |f| puts f.name }
```

Suppose you save this as `onrequest.rb`, you can use it by running `brew ruby onrequest.rb`.

**Reply by l008com** [reply]
This seems to give me more packages than the ones i manually installed. For example, freetds, libpq, libsodium, lynx etc. Those are definitely dependencies, I didn't install those. Also augeas, apr and some others.

**Reply by carlocab** [reply]
What's the output of
```
cat "$(brew --prefix freetds)/INSTALL_RECEIPT.json"
```
?

**Comment by l008com** [comment]
So it looks like the "installed_on_request" attribute doesn't work the intuitive way you would think. In other words, many packages that are installed, have this flag set to true even though they were still installed as dependencies. BUT it looks like if installed_on_request is true *AND* installed_as_dependency is false, then it's something I've manually installed. 
I'm using the `brew info --json` command and parsing the output in my own scripts (php) so hopefully this works....

**Reply by carlocab** [reply]
I'm not sure what you mean. I don't have any formulae installed that have `installed_on_request` as `true` but I never `brew install`ed or `brew reinstall`ed. That is, all my installed formulae that have `installed_on_request == true` are formulae that I `brew install`ed or `brew reinstall`ed at one point.

**Comment by l008com** [comment]
Ok so here's my script, where `$file` is a file containing the output of `brew info --json`:

```
$data = json_decode($file, true);
$result = array();
foreach ($data as $package)
	{
	if ( ($package['installed'][0]['installed_as_dependency'] != 1) and ($package['installed'][0]['installed_on_request'] == 1) )
		{ $result[] = $package['name']; }
	}
print_r($result);

```
Output:
```
Array
(
    [0] => certbot
    [1] => coreutils
    [2] => git
    [3] => imagemagick
    [4] => lynx
    [5] => mariadb
    [6] => openvpn
    [7] => php
)

```
Now that looks right_!_
